### PR TITLE
🩹 fix inferencers

### PR DESCRIPTION
**Pull request by ashwinvaidya17**
# Description

- Scores for classification models are unnormalized. We need to address that.

## Changes

- Adds support for classification models in inferencer.

## Checklist

- [x] My code follows the [pre-commit style and check guidelines](https://openvinotoolkit.github.io/anomalib/guides/using_pre_commit.html#pre-commit-hooks) of this project.
- [x] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing tests pass locally with my changes

# This test runs on nightly
image

### CODE REVIEW

**Review by samet-akcay — APPROVED** [review]

**Review by djdameln — COMMENTED** [review]
Isn't the torch inferencer supposed to run with the PyTorch model, and not the PL model? During inference we only need the raw predictions of the PyTorch model, and none of the training/metric/postprocessing stuff.

**Comment by samet-akcay** [comment]
I think we could start considering PL predict for torch inferencer and deprecate our torch interface

**Comment by ashwinvaidya17** [comment]
image
In the inferencer we create the lightning model object.

**Comment by djdameln** [comment]
> I think we could start considering PL predict for torch inferencer and deprecate our torch interface

That's also possible, and might be a good idea, because that's what the PL `predict` methods are meant for. But then do we need these changes? I'm afraid it would be confusing to add yet another method implementation to the PL modules. Also why only add this for these 3 models and not the others (padim, stfpm etc)?

**Comment by ashwinvaidya17** [comment]
> > I think we could start considering PL predict for torch inferencer and deprecate our torch interface
> 
> That's also possible, and might be a good idea, because that's what the PL `predict` methods are meant for. But then do we need these changes? I'm afraid it would be confusing to add yet another method implementation to the PL modules. Also why only add this for these 3 models and not the others (padim, stfpm etc)?

If we plan to switch then we don't need these changes. This was just to partially address the issues. Before this, inference was supported only for `padim, stfpm and patchcore`. These changes add support for the other models (except cflow).  This also extends the tests in [tests/nightly/deploy/test_inferencer.py](https://github.com/openvinotoolkit/anomalib/pull/105/files#diff-657ec31ec1761a238e937ccce2f9a390df8c99c89f6d3633eeb625d62e16251c)

**Comment by ashwinvaidya17** [comment]
> But how does it work for the other models? For padim for example, we don't implement the `forward` method. Why is it not needed in those models. It feels inconsistent to have this method implemented for some models, but not all.

Good point. In those models were the model name is `model`, the inferencer works well. This is because in the base `AnomalyModule` class we have a forward method which calls `self.model`. So this breaks for all the models which do not use the name `model` in the derived class. So this is a hacky way of overriding the forward method.

**Comment by ashwinvaidya17** [comment]
Closing as `inference.py` is going to be deprecated and most of the stuff here will be addressed in the OpenVINO PR.
